Gazebo: Moving platform

mbjd opened the pull request:
Gazebo plugin that controls a moving platform, producing random wobbling motion to simulate takeoff/landing on moving vehicles. 

Questions for reviewers: 
 - Only way to configure random motion frequency & amplitude is by changing the hardcoded constants. What would be a nicer way to do this (if needed)?
   - velocity now configurable, rest hardcoded still. 
 - Gazebo plugin depends on specific name of gazebo model, GZBridge callback depends on specific name of world. Is this acceptable? 
   - first one fixed, second one circumvented by not including PX4 communication in this PR. 

Launch using: 
```
PX4_GZ_MODEL_POSE=0,0,2.2,0,0,0 PX4_GZ_WORLD=moving_platform make px4_sitl gz_standard_vtol
```
with the desired model. Pose has to be adjusted so the vehicle is above the platform, which is at a height of 2m. By default the platform moves at 1 m/s -- this velocity can be changed with the environment variable `PX4_GZ_PLATFORM_VEL`:
```
PX4_GZ_PLATFORM_VEL=5 PX4_GZ_MODEL_POSE=0,0,2.2,0,0,0 PX4_GZ_WORLD=moving_platform make px4_sitl gz_standard_vtol
```


corresponding gazebo models merged [here](https://github.com/PX4/PX4-gazebo-models/pull/83).

dakejahl commented:
I think what you want is a system plugin and a model plugin. 
https://gazebosim.org/api/sim/8/createsystemplugins.html

The system plugin is added to the server.config
https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/simulation/gz_bridge/server.config#L16

The model plugin is added to the model (in the link below it's a sensor plugin, but same idea)
https://github.com/PX4/PX4-gazebo-models/blob/main/models/optical_flow/model.sdf#L72-L76

The system plugin is responsible for finding and registering the other plugin
https://github.com/PX4/PX4-Autopilot/blob/main/src/modules/simulation/gz_plugins/optical_flow/OpticalFlowSystem.cpp#L57-L92

You should probably then publish a TargetPose message in gz::transport. You can either create a new message like I did for optical flow (see gz_msgs/) or there might already exist in gz a suitable message. In GZBridge you would setup a callback for that topic (let's say LandingTargetPose) and then publish onto the uORB LandingTargetPose (add body velocities), or create a new uORB topic. This way the data interface is generic with respect to a "moving target". This way there could be multiple moving targets.

dakejahl commented:
Also I would recommend downloading the gazebo repositories and looking at the examples.
https://github.com/orgs/gazebosim/repositories?type=all

dakejahl commented:
> System/model plugin separation: Forgive the bluntness, but why? It seems to me that adding a system plugin (via server.config) in addition to the model plugin (via sdf) strictly adds complexity. If there is a concrete benefit that I'm currently missing, of course I'm happy to adapt.

Ah okay then I must have misunderstood how the system plugins work. I thought they only worked when added as world plugins. I  know the docs say otherwise but I couldn't figure out how to make my optical flow plugin work when added to the model directly rather than to the world. Great! This is actually better.

> Therefore we consider it preferable to handle this communication with an external tool rather than going through PX4 and "polluting" upstream.

I think a moving target is a great addition to upstream, but it takes slightly more effort to make it generic. Ultimately it's up to you. I would create a custom proto (let's say `LandingTarget`) which contains the same GPS lat/lon/vel information as NavSat, but the message also includes an ID. This way the GZBridge can regex subscribe to all publishers of this new message. This way you avoid needing to know the model name to subscribe to the NavSat msg, and you can support multiple mvoing platforms. The lat/lon/vel you publish could come directly from the model pose, or you could add the NavSat sensor to your model sdf subscribe to the NavSat callback within your plugin (I think you can derive the topic path from the model sdf).

mbjd commented:
Great to hear about the plugin architecture :) 

As for the rest, I think we'll proceed for now without making a new proto message and interfacing to PX4, and rely just on the navsat and IMU plugin to get info out from the platform. In the future we can make this generic and extensible by publishing the info on a topic independent of the specific SDF structure and reading it in GZBridge as you describe, but I'd like to keep it separate from this PR.

mbjd commented:
Reabsed & marked ready for review as now this does not depend on the specific implementation in the gz submodule anymore :)

dakejahl commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.hpp:
```diff
@@ -0,0 +1,123 @@
+ *        </model>
+ *      </sdf>
+ *
+ */
```
maybe this comment and the mention of the env var can go in a README.md in this plugin dir? I don't know... might be helpful to have info consolidated

dakejahl reviewed (commented):
you checked in gazebo-classic changes but otherwise LGTM

mbjd commented:
Added a README, removed template comments, removed erroneous gazebo classic changes (thanks for pointing out!) 

Also, small mod so it doesn't start at full velocity but ramps up at constant acceleration. This makes it possible to choose relatively fast velocities without the aircraft sliding off the platform immediately.

Jaeyoung-Lim commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,197 @@
+	// write these in terms of time constants?
+	// https://ethz.ch/content/dam/ethz/special-interest/mavt/dynamic-systems-n-control/idsc-dam/Lectures/Signals-and-Systems/Lectures/Lecture%20Notes%209.pdf
+	const double alpha_v = 0.01;
+	const double alpha_w = 0.01;
```
what is alpha? Could you briefly add in the comments what this means?

mbjd replied:
Coefficients for the low-pass filters. I renamed them `filter_coef_*` and changed the comment to use that terminology as well to hopefully make it more obvious.

Jaeyoung-Lim commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,197 @@
+						    _platform_orientation.Z()
+					    );
+
+		_platform_w += -attitude_gain * sgn * q_imag;
```
Why are we controlling the platform movement with angular velocity? Can't you just constrain the movement?

mbjd replied:
Any further thoughts on this or should I mark this as resolved?

Jaeyoung-Lim replied:
I see now what problem you were trying to solve. However, if you are concerned about the contact dynamics, using body rate commands would still introduce the same problems as you set exact pose, as this imposes a kinematics constraint on the simulation. Therefore, using wrench commands would have been the correct way to solve this.

mbjd replied:
Yes, you're right that technically the same problem with the kinematic constraint persists. However, practically it is now so small as to be negligible. All of our use cases have platform mass >> drone mass, making the kinematic constraint a close approximation of reality. 

Do you consider a second-order, wrench-based implementation a must-have for this PR? (either to account for use cases where drone and platform mass are on similar scales, or as a matter of principle)? Or is your comment more of a suggestion for future improvement?

I'm open to changing it, but from our side it's not a requirement.

mbjd replied:
Okay, makes sense. I will change accordingly early next week, leaving this PR open until then. 

ATM we're not blocked, for testing we just merge this branch with others as necessary :)

mbjd replied:
Adressed this in the newest commits, could you re-review? thanks :)

Jaeyoung-Lim reviewed (changes requested):
- The commit history needs to be cleaned up before being able to merge

mbjd commented:
@Jaeyoung-Lim about commit history, yes, definitely agree. feel free to squash when merging, or I can squash manually if you prefer.

RomanBapst commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,198 @@
+	_entity = entity;
+	_model = gz::sim::Model(entity);
+
+	double default_mean_vel = 1.0f;
```
@mbjd Just remove this and set the other variable directly

Jaeyoung-Lim commented:
@mbjd Please write proper commit messages such that the changes are referred properly: https://community.atlassian.com/forums/Bitbucket-articles/Git-Best-Practices/ba-p/1628803#:~:text=Write%20Good%20Commit%20Messages,by%20including%20a%20blank%20line.

RomanBapst commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,198 @@
+	// we can't really simulate fast velocities as the drone just slips off
+	// the platform.
+
+	double max_accel_mag = 2.; // m/s^2
```
```suggestion
	const double max_accel_mag = 2.; // m/s^2
```

RomanBapst commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,198 @@
+	// the platform.
+
+	double max_accel_mag = 2.; // m/s^2
+	double dt = std::chrono::duration<double>(_info.dt).count(); // s
```
```suggestion
	const double dt = std::chrono::duration<double>(_info.dt).count(); // s
```

Jaeyoung-Lim commented:
> Github let's you squash and merge and add a new commit message at the time of merging. I would actually prefer that since it preserves history during development. When someone rebases you can't tell what's changed and you have to review the entire file again to be 100% sure things didn't regress. My 2 cents

@dakejahl I agree with your comments, but on the basis that the commit history is already inline with the best practices. This is just my opinion.
- If the commit histories are not organized properly (e.g. including messages like "Its working!", as was in this PR), the history during development is not worth preserving IMHO. I would even go on and say that if the commit messages are meaningful, it is better to do a rebase and merge than a squash. 
- If one of the maintainers are to merge the PR, I don't think it is so nice to rewrite the intent of the commit. I think this should come from the author, which should be written on the basis of the review comment the person received during the PR process.

dakejahl commented:
> I don't think it is so nice to rewrite the intent of the commit. I think this should come from the author

I agree with this for sure.

The reason I don't like rebasing is entirely because of the loss of history. When someone rebases it's not possible to know if they slipped in some malicious code unless you re-review the entire PR. For large PRs this is unrealistic. I'm thinking about this from a zero trust perspective. Whereas a single commit I can see exactly what the author changed inbetween reviews.

> I would even go on and say that if the commit messages are meaningful, it is better to do a rebase and merge than a squash.

Yeah... maybe. IMO every commit in main should build. Otherwise bisecting is impossible. During development it's a lot of extra work to have good multi-commits. At least for me, when I'm working I commit frequently and the commits are often not worthy of a good description and may not build. The commit is more about saving your work.

Anyway. This PR looks good to me! Merge when ready.

mbjd commented:
Update: changed to wrench implementation as recommended [here](https://github.com/PX4/PX4-Autopilot/pull/24471#discussion_r1998634343). Now it works purely using forces and torques rather than setting velocities directly. 

Other updates:
 - Also make heading configurable by env var in addition to velocity
 - Cleaner error handling
 - Read parameters (gravity, platform mass & height) from model rather than hardcoding
 - Update README with new env vars, usage in sdf, etc.



@Jaeyoung-Lim @dakejahl please feel free to review in as much detail as you deem necessary after you've already reviewed the previous version. Currently this still points the `Tools/simulation/gz` submodule to a [PR branch](https://github.com/PX4/PX4-gazebo-models/pull/95). Please merge only after that PR is merged and I've updated this one accordingly.

hamishwillee commented:
@mbjd When this is done, would be great to have docs (hence the tag addition). This would be an addition to the world of course, but I suspect more would be needed. I'm not sure what that would be since I am not a domain expert, but enough so that someone can take all this cool work you have done and use it in a turnkey way. Perhaps a dedicated doc or section on simulation of landing on moving targets.

perrrre commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,276 @@
+		if (pose_component) {
+			_platform_height_setpoint = pose_component->Data().Z();
+		} else {
+			gzwarn << "Unable to get inertial component for link " << link_name << ". Keeping default mass of " << _platform_mass << std::endl;
```
```suggestion
			gzwarn << "Unable to get pose component for link " << link_name << ". Keeping default height set point of " << _platform_height_setpoint << std::endl;
```

mbjd replied:
oops, thanks for pointing that out, fixed!

perrrre commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,276 @@
+	const double noise_ampl_force = 1. * _platform_mass;
+	const double noise_ampl_torque = 1. * _platform_mass;
+
+	_noise_lowpass_force = (1 - filter_coef_force) * _noise_lowpass_force + filter_coef_force * noise_force;
```
you could use this: PX4-Autopilot/src/lib/mathlib/math/filter/AlphaFilter.hpp

mbjd replied:
Yes, thought about that briefly too. I chose not to because I remember some discussion about potentially moving the gazebo plugins into the PX4-gazebo-models repo, in which case I'm not sure if we can and should have it rely on PX4 stuff.

perrrre replied:
make sense!

mbjd commented:
Rebased on main again & removed commits switching gz submodule branches to avoid merge conflicts. 

@hamishwillee yes, definitely. the current [readme](https://github.com/PX4/PX4-Autopilot/blob/7f3cfeca4f5eb9c02b03dde33376d7f62e930974/src/modules/simulation/gz_plugins/moving_platform_controller/README.md) provides a basic overview but is hardly discoverable to users. Would this ["argument" list](https://docs.px4.io/main/en/sim_gazebo_gz/#syntax) be a suitable place to add it in your opinion, or would you prefer adding an extra section? Because it is more of a special case of the existing PX4_GZ_WORLD rather than just another environment variable...

perrrre commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,276 @@
+	// Feedback terms to ensure stability of the platform.
+	{
+
+		const gz::math::Vector3d pos_gains = _platform_mass * gz::math::Vector3d(0., 0., 1.);
```
would be nice with units to simplify the reading

mbjd replied:
addressed in latest commit :)

KonradRudin commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,324 @@
+	updateNoise(dt_sec);
+
+	// Wait for model to spawn, takes quite some time now...
+	const bool keep_stationary = _startup_timer.ElapsedTime() < 5s;
```
Can you quickly explain why you need to keep it stationary for 5s?

mbjd replied:
because the model takes 5s to spawn. If we don't wait for it, the platform moves away before it spawns.

Jaeyoung-Lim replied:
@mbjd Doing this based on time would not be reliable, as this would largely depend on the system. You can check if the entity exists through the gz API.

mbjd replied:
@Jaeyoung-Lim true that's a better solution, done in latest commit.

KonradRudin commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,324 @@
+	const double filter_coef_force = dt / (dt + time_constant_force);
+	const double filter_coef_torque = dt / (dt + time_constant_torque);
+
+	_noise_lowpass_force = (1. - filter_coef_force) * _noise_lowpass_force + filter_coef_force * noise_force;
```
Nitpicking. This is more an aceleration and not a force, right?

mbjd replied:
can also be thought of that way - I presume you mean because of the scaling by mass and moment of inertia. the way I imagined it: 
 - `_noise_lowpass_force` and `_noise_lowpass_torque` are dimenionless
 - `_force` is an actual force in N, `_torque` is an actual torque in Nm
 - For this to work, `noise_ampl_force` has to be in N and `noise_ampl_torque` in Nm. Compared to them being scaled by mass and moment of inertia respectively, we have to imagine multiplying in the right unit conversion factors. 
 
If you want I can make this explicit and e.g. calculate linear/angular accelerations and only in the end convert to forces.

KonradRudin reviewed (commented):
Just a question, otherwise LGTM

Jaeyoung-Lim commented on src/modules/simulation/gz_plugins/moving_platform_controller/MovingPlatformController.cpp:
```diff
@@ -0,0 +1,325 @@
+		const auto gravity = world.Gravity(ecm);
+
+		if (gravity.has_value()) {
+			_gravity = gravity.value().Z();
```
You can also disable gravity for this link?

Jaeyoung-Lim replied:
I think it is good enough to go in at this point, I just think there are ways to improve it
